### PR TITLE
Add example app to configure EJBs over two-way TLS-encrypted HTTP connections

### PR DESCRIPTION
**Pull request by cam-rod**

### CODE REVIEW

File: ejb-https/README.md [hunk 1]
```diff
@@ -0,0 +1,272 @@
+# ejb-https: Configuring Jakarta Enterprise Beans over HTTP with TLS
```
Comment (fjuma) [inline ejb-https/README.md:1]: Maybe we could reword this slightly to something like "Accessing Jakarta Enterprise Beans with mutual TLS over HTTPS". What do you think?

File: ejb-https/README.md [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+The `ejb-https` example demonstrates how to securely connect to EJBs using HTTP with two-way TLS
```
Comment (fjuma) [inline ejb-https/README.md:3]: Same here, i.e., using mutual TLS over HTTPS.

File: ejb-https/README.md [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+encryption, and control access to EJBs with a SCRAM-SHA-512-PLUS SASL authentication mechanism.
```
Comment (fjuma) [inline ejb-https/README.md:4]: s/with a/with the

File: ejb-https/README.md [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+Note: The purpose of this example is to demonstrate more advanced security configurations using the
+Elytron subsystem. If you haven't already, take a look at the following examples:
+- [How to connect to an EJB using HTTP and basic authentication](https://github.com/wildfly-security-incubator/elytron-examples/tree/main/ejb-http)
+- [How to connect to an EJB using SASL authentication](https://github.com/wildfly-security-incubator/elytron-examples/tree/main/ejb-security)
+
+### Overview
+
+This quickstart takes the following steps to implement EJB security using Elytron:
+
+1. **Add a secured identity for accessing the EJB**
```
Comment (fjuma) [inline ejb-https/README.md:15]: I think we could reword this to "Add an identity for accessing a secured EJB"

File: ejb-https/README.md [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+   1. Create a filesystem-based identity store to be used by the Elytron subsystem.
+   2. Add an `sslApp` security domain mapping in the `ejb3` subsystem to enable Elytron security for the `SecuredEJB`.
+2. **Create an SASL authentication mechanism**
```
Comment (fjuma) [inline ejb-https/README.md:18]: I think we could reword this to "Configure an additional SASL authentication mechanism"

File: ejb-https/README.md [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+   1. Create a sasl-authentication-factory used by the Elytron subsystem to offer SCRAM-SHA-512-PLUS as an authentication mechanisms.
```
Comment (fjuma) [inline ejb-https/README.md:19]: `sasl-authentication-factory`

Comment (fjuma) [inline ejb-https/README.md:19]: s/an authentication mechanisms/an authentication mechanism

File: ejb-https/README.md [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+   2. Configure a credential store on the client side to be used for the authentication mechanism.
```
Comment (fjuma) [inline ejb-https/README.md:20]: Since this credential store is used on the client side, I think this can be pulled out into a separate section since the SASL mechanism configuration described here is for the server side.

File: ejb-https/README.md [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+3. **Generate certificates for TLS-encrypted connections**
+   1. Generate a public-private key pair on both the client and server side, and secure each pair in a key store.
```
Comment (fjuma) [inline ejb-https/README.md:22]: s/secure each key pair/store each key pair

File: ejb-https/README.md [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+   2. Export self-signed certificates from both sides, and import opposite certificates into a trust store.
+   3. Configure the `https-listener` to use the key store and trust store when performing a TLS handshake.
```
Comment (fjuma) [inline ejb-https/README.md:24]: Just a general comment. On the client side, there's also TLS related configuration added in the wildfly-config.xml file. I know this section here is just meant to be a quick summary. I think it would be good either in the blog post or elsewhere in the README to make sure we clearly distinguish between sever side configuration and client side configuration.

File: ejb-https/README.md [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+4. **Configure the EJB to use TLS and SASL when being accessed**
```
Comment (fjuma) [inline ejb-https/README.md:25]: I think this could be reworded to something like "Secure the EJB using TLS and SASL"

File: ejb-https/README.md [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+   1. Configure a `http-connector` to use both the `https-listener` and SASL mechanism ???
```
Comment (fjuma) [inline ejb-https/README.md:26]: Right, you'll want to configure the `sasl-authentication-factory` as well.

File: ejb-https/README.md [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+First, clone the `elytron-examples` repo locally:
+```shell
+    $ git clone https://github.com/wildfly-security-incubator/elytron-examples
+    $ cd elytron-examples
+```
+Next, start the server:
+
+```shell
+    $ WILDFLY_HOME/bin/standalone.sh
+```
```
Comment (fjuma) [inline ejb-https/README.md:53]: Just to check, should the above two steps be in the credential store section or should they be pulled out?

File: ejb-https/README.md [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+    $ keytool -genkeypair -alias client -keyalg RSA -keysize 2048 -validity 365 -keystore /PATH/TO/sslClient.keystore.jks -dname "CN=client" -keypass clientKeyPairSecret -storepass clientKeySecret
+```
+
+Export a self-signed certificate and save it locally to a certificate file:
+```shell
+    $ keytool -exportcert -keystore /PATH/TO/sslClient.keystore.jks -alias client -keypass clientKeyPairSecret -storepass clientKeySecret -file /PATH/TO/sslClient.cer
+```
+
+_\(Note: self-signed certificates are not suitable for production environments, and treated as insecure by
+most programs. To obtain a certificate signed by a Certificate Authority, take a look at this [blog post](https://wildfly-security.github.io/wildfly-elytron/blog/obtaining-certificates-from-lets-encrypt-using-the-wildfly-cli/).\)_
```
Comment (fjuma) [inline ejb-https/README.md:89]: Really like that you included this here!

File: ejb-https/README.md [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+You should see the following result when you run the script:
+```shell
+    The batch executed successfully
+    The batch executed successfully
+    The batch executed successfully
+    The batch executed successfully
+    process-state: reload-required
+```
+
+### Import the Server Certificate and Client Stores
```
Comment (fjuma) [inline ejb-https/README.md:123]: Could reword "Client Stores" to "Client TLS configuration" or something along those lines.

File: ejb-https/README.md (hunk 9 above)
Comment (fjuma) [reply ejb-https/README.md:24]: I just got further down and do see the TLS related configuration for the client side below, thanks!

File: ejb-https/README.md [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+
+### Build and Deploy the Demo App
+1. Make sure you start the WildFly server as described above.
+2. Open a new terminal and navigate to the root directory of this example.
+3. Type the following command to build the artifacts.
+```shell
+    $ mvn clean install wildfly:deploy
+```
+
+This deploys the `{artifactId}/target/{artifactId}.jar` to the running instance of the server.
```
Comment (fjuma) [inline ejb-https/README.md:154]: The placeholders here should be updated.

File: ejb-https/configure-elytron.cli [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+/subsystem=elytron/filesystem-realm=sslFsRealm:add-identity(identity=example_user)
+
+# Add the password for the example identity
+/subsystem=elytron/filesystem-realm=sslFsRealm:set-password(identity=example_user, clear={password=examplePwd1!})
+
+# Add the guest role for the example identity
+/subsystem=elytron/filesystem-realm=sslFsRealm:add-identity-attribute(identity=example_user, name=Roles, value=[guest])
+
+# Add a simple role decoder
+/subsystem=elytron/simple-role-decoder=from-roles-attribute:add(attribute=Roles)
```
Comment (fjuma) [inline ejb-https/configure-elytron.cli:21]: Because the default role decoder reads from the "Roles" attribute, I think you can remove this.

**Review by fjuma — COMMENTED** [review]
@cam-rod Thanks very much for the PR! This looks excellent! I've added some small comments and will also try this out locally.

File: ejb-https/README.md [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+    $ WILDFLY_HOME/bin/elytron-tool.sh credential-store --location "/PATH/TO/mycredstore.cs" --password clientStorePassword --add example_user --secret examplePwd1!
+```
+
+In `wildfly-config.xml`, make sure to change the path to your credential store (i.e. replace `/PATH/TO/` with
+the path to your credential store):
+
+```xml
+    <attributes>
+        <attribute name="keyStoreType" value="JCEKS"/>
+        <attribute name="location" value="/PATH/TO/sslCredStore.cs"/>
```
Comment (fjuma) [inline ejb-https/README.md:72]: Just a small thing I noticed - the commands above use mycredstore.cs but the wildfly-config.xml file uses sslCredStore.cs. Would be good to update this so one name is used.

File: ejb-https/README.md [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+    </attributes>
+```
+
+### Generate the Client Key Pair and Certificate
+
+Generate the client's key pair, and store it locally in a Java KeyStore as follows:
+```shell
+    $ keytool -genkeypair -alias client -keyalg RSA -keysize 2048 -validity 365 -keystore /PATH/TO/sslClient.keystore.jks -dname "CN=client" -keypass clientKeyPairSecret -storepass clientKeySecret
```
Comment (fjuma) [inline ejb-https/README.md:80]: To simplify things a bit, I think we could remove the `keypass` and just set `storepass=clientSecret`.

File: ejb-https/README.md [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+```
+
+Export a self-signed certificate and save it locally to a certificate file:
```
Comment (fjuma) [inline ejb-https/README.md:83]: s/Export a/Export the

File: ejb-https/README.md (hunk 11 above)
Comment (cam-rod) [reply ejb-https/README.md:26]: oh i completely forgot to change that, whoops

File: ejb-https/README.md (hunk 12 above)
Comment (cam-rod) [reply ejb-https/README.md:53]: probably would be better to pull out, changed

File: ejb-https/configure-elytron.cli [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+
+
+## SSL, SASL and Connector Configuration ##
+# Add the server SSL context to reference the trust and key managers. This context requires client authentication for two-way SSL, and supports TLSv1.2 and TLSv1.3
+/subsystem=elytron/server-ssl-context=sslSSC:add(key-manager=sslKM,protocols=["TLSv1.3","TLSv1.2"],trust-manager=sslTM,need-client-auth=true)
+
+# Run the batched commands
+run-batch
+
+# Save the current legacy security realm or SSL context of the HTTPS listener to file for later restoration
```
Comment (fjuma) [inline ejb-https/configure-elytron.cli:78]: This shouldn't be needed (see my comment below).

File: ejb-https/README.md (hunk 18 above)
Comment (fjuma) [reply ejb-https/README.md:80]: I think we should also drop the .jks suffix since I believe the keystore type is now PCKS12 by default.

File: ejb-https/configure-elytron.cli (hunk 20 above)
Comment (cam-rod) [reply ejb-https/configure-elytron.cli:78]: also removed related section from the README

**Comment by fjuma** [comment]
@Skyllarr Whenever you get a chance later this week, please review and try this out as well. Thanks.

**Comment by cam-rod** [comment]
Modified `server-ssl-context` to enable TLS 1.3 (was implied but `cipher-suite-names` attribute was not set).

File: ejb-https/README.md (hunk 1 above)
Comment (fjuma) [reply ejb-https/README.md]: @cam-rod I just realized that this title might be confusing because we also have a separate feature known as "EJB over HTTP" (as described here: https://wildfly-security.github.io/wildfly-elytron/blog/ejb-over-http/#performing-invocations).

To prevent others from thinking that's somehow related here, what if we rename this one more time to:

"Accessing Jakarta Enterprise Beans over mutual TLS"

File: ejb-https/README.md [hunk 21]
```diff
@@ -0,0 +1,238 @@
+# ejb-https: Accessing Jakarta Enterprise Beans with HTTP over mutual TLS
+
+The `ejb-https` example demonstrates how to securely connect to EJBs using HTTP over mutual TLS
```
Comment (fjuma) [inline ejb-https/README.md:3]: Same here, maybe this could be something like: "connect to EJBs deployed to WildFly with mutual TLS"

File: ejb-https/README.md (hunk 1 above)
Comment (cam-rod) [reply ejb-https/README.md]: sounds good

**Comment by fjuma** [comment]
Thanks for the updates, @cam-rod! Whenever you get a chance, the last two commits can be squashed together. Let us know if you need any help with that.

**Review by fjuma — APPROVED** [review]
Thanks @cam-rod!

**Review by Skyllarr — APPROVED** [review]

**Comment by Skyllarr** [comment]
@cam-rod Thanks!
